src/components/CountryPicker/types.ts

```diff
@@ -0,0 +1,67 @@
+import { countries, states } from './constants';
+
+// Types
+
+export type PrependTrigger<T> = {
+  [K in keyof T as K extends 'hasError' ? 'hasTriggerError' : `trigger${Capitalize<string & K>}`]: T[K];
+};
+
+export type ModalStateProps = {
+  /**
+   * Determines whether the modal is open.
+   */
+  isOpen?: boolean;
+
+  /**
+   * Callback function triggered when the modal is closed.
+   */
+  onClose?: () => void;
+};
+
+export type CommonProps = {
+  /**
+   * The base class name for styling the component.
+   */
+  baseClassName?: string;
+
+  /**
+   * Configuration for phone/country variant.
+   */
+  variantConfig: VariantConfig;
+};
+
+export type ModalBaseProps<T> = T & CommonProps & ModalStateProps;
+
+export type Country = {
+  name: (typeof countries)[number]['name'];
+  code: (typeof countries)[number]['code'];
+};
+
+export type State = {
+  name: (typeof states)[number]['name'];
+  code: (typeof states)[number]['code'];
+};
+
+export type CountryCallingCode = `+${number}`;
+
+export type PhoneConfig = {
+  isPhone: true;
+  countryValue?: Country['code'];
+  onChange: (value: Country['code']) => void;
+};
+
+export type CountryConfig = {
+  isPhone: false;
+  countryValue?: Country['name'];
+  onChange: (value: Country['name']) => void;
+};
+
+export type VariantConfig = PhoneConfig | CountryConfig;
```

Comment:
this is a really interesting approach.  From a consumer perspective is might be a little more complex to consume rather than a simple `variant='phone' | 'country'` prop and having the "value" and "onChange" be at the top-level prop with one or the other values or onChange.  The only other thing I can think of is that if we plug this into a remix-form, it sort of expects `value` and `onChange` to be top-level props to automatically register and plug them into the form draft state.  You might play with that and see how well it plays with this variantContif approach